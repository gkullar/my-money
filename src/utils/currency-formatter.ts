export const toGBP = (value: number): string => 'Â£' + (value / 100).toFixed(2);
