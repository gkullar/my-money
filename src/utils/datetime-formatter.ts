export const toDateTime = (value: string): string =>
  new Date(value).toLocaleString();
