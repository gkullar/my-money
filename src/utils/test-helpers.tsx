import { ReactElement } from 'react';
import { render, RenderOptions } from '@testing-library/react';
import 'jest-styled-components';
import AppProviders from '../app-providers';

const customRender = (ui: ReactElement, options?: RenderOptions) =>
  render(ui, { wrapper: AppProviders, ...options });

export * from '@testing-library/react';
export { customRender as render };
