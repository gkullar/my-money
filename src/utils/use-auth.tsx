import React, {
  createContext,
  useContext,
  useState,
  FunctionComponent
} from 'react';
import uuid from './uuid';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;
const stateTokenKey = `${process.env.REACT_APP_API_STATE_TOKEN_KEY}`;
const clientId = `${process.env.REACT_APP_API_CLIENT_ID}`;
const clientSecret = `${process.env.REACT_APP_API_CLIENT_SECRET}`;
const redirectUri = 'http://localhost:3000'; // app not for production!

if (!localStorage.getItem(stateTokenKey))
  localStorage.setItem(stateTokenKey, uuid());

const stateToken = localStorage.getItem(stateTokenKey);

interface State {
  isAuthenticated: boolean;
  login: () => void;
  logout: () => void;
}

const authContext = createContext<State>({} as any);

const useAuth = () => useContext(authContext);

const AuthProvider: FunctionComponent<{}> = ({ children }) => {
  const auth = useAuthProvider();
  return <authContext.Provider value={auth}>{children}</authContext.Provider>;
};

function useAuthProvider(): State {
  const [isAuthenticated, setIsAuthenticated] = useState<boolean>(
    localStorage.getItem(accessTokenKey) ? true : false
  );
  const params = new URLSearchParams(window.location.search);

  if (!isAuthenticated && params.has('code') && params.has('state')) {
    if (params.get('state') !== stateToken)
      throw new Error('Error: state token not matched'); // @todo handle error in UI

    authenticate(params.get('code') as string);
  }

  async function authenticate(code: string) {
    const data = new FormData();
    data.append('grant_type', 'authorization_code');
    data.append('client_id', clientId);
    data.append('client_secret', clientSecret);
    data.append('redirect_uri', redirectUri);
    data.append('code', code);

    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/token`,
      {
        method: 'post',
        body: data
      }
    );

    const result = await response.json();

    if (response.status !== 200 || result.error)
      throw new Error('Error: Failed Authentication'); // @todo handle error in UI

    localStorage.setItem(accessTokenKey, result.access_token);
    setIsAuthenticated(true);
    window.history.replaceState(null, document.title, window.location.pathname); // remove query params
  }

  function login() {
    window.location.href = `https://auth.monzo.com/?client_id=${clientId}&redirect_uri=${redirectUri}&response_type=code&state=${stateToken}`;
  }

  async function logout() {
    const response = await fetch(
      `${process.env.REACT_APP_API_URL}/oauth2/logout`,
      {
        method: 'post',
        headers: new Headers({
          Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
        })
      }
    );

    const result = await response.json();

    if (!result.ok) throw new Error('Error: Failed to logout'); // @todo handle error in UI

    localStorage.removeItem(accessTokenKey);
    setIsAuthenticated(false);
  }

  return { isAuthenticated, login, logout };
}

export { useAuth, AuthProvider };
