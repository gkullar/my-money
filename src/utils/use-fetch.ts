import { useState, useEffect } from 'react';

const accessTokenKey = `${process.env.REACT_APP_API_ACCESS_TOKEN_KEY}`;

interface State<T> {
  data: T;
  loading: boolean;
  hasErrors: boolean;
  hasPermissions: boolean;
}

const useFetch = <T>(url: string): State<T> => {
  const [data, setData] = useState<T>({} as T);
  const [loading, setLoading] = useState<boolean>(true);
  const [hasErrors, setHasErrors] = useState<boolean>(false);
  const [hasPermissions, setHasPermissions] = useState<boolean>(true);

  async function fetchUrl() {
    const response = await fetch(url, {
      method: 'get',
      headers: new Headers({
        Authorization: `Bearer ${localStorage.getItem(accessTokenKey)}`
      })
    });

    const result = await response.json();

    if (response.status === 401) {
      localStorage.removeItem(accessTokenKey);
      window.location.href = '/';
    } else if (response.status === 403) {
      setHasPermissions(false);
    } else if (response.status !== 200) {
      throw new Error('Error: Failed Data Fetch'); // @todo handle error in UI
    } else {
      setData(result);
      setLoading(false);
      setHasErrors(false);
      setHasPermissions(true);
    }
  }

  useEffect(() => {
    fetchUrl();
  }, []);

  return { data, loading, hasErrors, hasPermissions };
};

export default useFetch;
