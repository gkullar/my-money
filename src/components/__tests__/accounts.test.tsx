import React from 'react';
import Accounts from '../accounts';
import { render, wait } from '../../utils/test-helpers';

describe('Accounts', () => {
  test('should render an error message if error', async () => {
    const spy = jest.spyOn(window, 'fetch') as jest.Mock<any>;
    spy.mockResolvedValue({ status: 404 });

    const { container } = render(<Accounts />);

    await wait(() =>
      expect(container.textContent).toContain(
        'There was an error retrieving accounts data'
      )
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  test('should render accounts if no error', async () => {
    const spy = jest.spyOn(window, 'fetch') as jest.Mock<any>;
    spy.mockResolvedValue({
      status: 200,
      json: () =>
        Promise.resolve({
          accounts: [
            {
              id: 'TestAccountIs',
              account_number: 'TestAccountNo',
              description: 'TestAccountDescription',
              created: 'TestAccountCreated'
            }
          ]
        })
    });

    const { container } = render(<Accounts />);

    await wait(() => expect(container.textContent).toContain('TestAccountNo'));

    expect(container.firstChild).toMatchSnapshot();
  });
});
