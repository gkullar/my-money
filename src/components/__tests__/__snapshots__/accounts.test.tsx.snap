// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Accounts should render accounts if no error 1`] = `
.c15 {
  padding: 20px;
  position: relative;
  border-radius: 15px;
  color: #1b1b1b;
  background: rgba(255,255,255,0.9);
  box-shadow: 8px 8px 4px -4px rgba(55,65,79,0.4);
}

.c3 {
  padding: 20px;
  position: relative;
  border-radius: 15px;
  color: #fff;
  background: rgba(0,219,163,0.9);
  box-shadow: 8px 8px 4px -4px rgba(95,107,125,0.4);
}

.c2 {
  width: 310px;
  height: 225px;
}

.c6 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
  margin-right: 15px;
}

.c9 {
  font-size: 0.75rem;
  text-transform: uppercase;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  margin-bottom: 5px;
}

.c7 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  display: inline-block;
}

.c10 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
  font-weight: bold;
  font-size: 1.5rem;
}

.c13 {
  font-size: 1.3rem;
  -webkit-letter-spacing: 1px;
  -moz-letter-spacing: 1px;
  -ms-letter-spacing: 1px;
  letter-spacing: 1px;
}

.c4 {
  padding: 0px 5px;
}

.c5 {
  margin-top: 5px;
  margin-bottom: 30px;
}

.c8 {
  border-bottom: 1px solid #fff;
  padding-bottom: 20px;
  margin-bottom: 20px;
}

.c11 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c12 {
  width: 50%;
}

.c14 {
  width: 310px;
  height: 225px;
  background: rgba(0,0,0,0.1);
  box-shadow: none;
  border: 2px dashed rgba(0,0,0,0.3);
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  height: 100%;
}

.c1 {
  margin-bottom: 20px;
}

@media (min-width:30em) {
  .c2 {
    width: 350px;
  }
}

@media (min-width:30em) {
  .c14 {
    width: 350px;
  }
}

@media (min-width:30em) {
  .c1 {
    margin-right: 20px;
  }
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <div
          class="c5"
        >
          <div
            class="account-detail"
          >
            <div
              class="c6"
            >
              Account No
            </div>
            <div
              class="c7"
            >
              TestAccountNo
            </div>
          </div>
        </div>
        <div
          class="c8"
        >
          <div
            class="account-detail"
          >
            <div
              class="c9"
            >
              Balance
            </div>
            <div
              class="c10"
            >
              £NaN
            </div>
          </div>
        </div>
        <div
          class="c11"
        >
          <div
            class="c12"
          >
            <div
              class="account-detail"
            >
              <div
                class="c9"
              >
                Available
              </div>
              <div
                class="c13"
              >
                £NaN
              </div>
            </div>
          </div>
          <div
            class="c12"
          >
            <div
              class="account-detail"
            >
              <div
                class="c9"
              >
                Todays Spend
              </div>
              <div
                class="c13"
              >
                £NaN
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    class="c1"
  >
    <div
      class="c14 c15"
    />
  </div>
  <div
    class="c1"
  >
    <div
      class="c14 c15"
    />
  </div>
</div>
`;

exports[`Accounts should render an error message if error 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-flex-wrap: wrap;
  -ms-flex-wrap: wrap;
  flex-wrap: wrap;
  height: 100%;
}

<div
  class="c0"
>
  <p>
    There was an error retrieving accounts data
  </p>
</div>
`;
